### PR TITLE
Add Grafana Beyla Flow component

### PR DESCRIPTION
**Pull request by marctc**
#### PR Description

This PR adds a wrapper for [Beyla](https://github.com/grafana/beyla), which is eBPF collector to automatically instrument applications.

#### PR Checklist

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [ ] Config converters updated

### CODE REVIEW

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (grcevski) [inline docs/sources/flow/reference/components/beyla.ebpf.md:14]: Do we want to explicitly mention that OTel metrics are not supported at the moment? It might be easy to overlook...

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (grcevski) [inline docs/sources/flow/reference/components/beyla.ebpf.md:17]: We also need CAP_SYS_PTRACE and when running in k8s we should mention that app armour must be disabled for the deployment, e.g:

```
container.apparmor.security.beta.kubernetes.io/beyla: "unconfined" 
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (mariomac) [inline docs/sources/flow/reference/components/beyla.ebpf.md:12]: Shall we mention that this component might be some versions behind the `latest` version of Beyla?

Comment (marctc) [reply docs/sources/flow/reference/components/beyla.ebpf.md:12]: I think is usually not necesary specify the version of that a component uses. @ptodev do you have any insights here? We probably should mark as beta, what do you think?

Comment (mariomac) [reply docs/sources/flow/reference/components/beyla.ebpf.md:12]: I'm thinking in the time between we publish a new Beyla version with new public features and we later update the Beyla library here (and publish a new version of the Agent)... maybe a user would expect that a given feature is working already in the Agent.

File: internal/component/beyla/args.go
Comment (rfratto) [inline internal/component/beyla/args.go:1]: To align with other components, where the name of the component matches the folder structure, this should be in `internal/component/beyla/ebpf/args.go`

File: internal/component/beyla/beyla_linux.go
Comment (rfratto) [inline internal/component/beyla/beyla_linux.go:144]: This will cause a deadlock as the mutex will continue to be held on exit. Can you move `c.mut.Unlock()` to a defer?

File: internal/component/beyla/beyla_linux.go
Comment (rfratto) [inline internal/component/beyla/beyla_linux.go:130]: I don't recommend returning an error from Run; this would be considered a fatal error and the component will not be restarted. 

Instead, consider replacing this with a log line, a `continue`, and setting the component health to Unhealthy to signal to the user that there's something wrong with the arguments.

Comment (rfratto) [reply internal/component/beyla/beyla_linux.go:130]: (Also the mutex doesn't get released after this returns)

File: internal/component/beyla/beyla_linux.go
Comment (rfratto) [inline internal/component/beyla/beyla_linux.go:175]: Can you help me understand what the `/metrics` endpoint will return metrics for? I read the docs but it wasn't immediately clear. 

Metrics specific to Beyla itself (and not any applications it is monitoring) should be registered to the registry given to the component. 

However, metrics specific to applications Beyla is monitoring should be done via this pattern, where it exposes those metrics separately and not over the default metrics endpoint.

File: internal/component/beyla/beyla_linux.go
Comment (rfratto) [inline internal/component/beyla/beyla_linux.go:179]: This is taking the address of the default config from the Beyla package, and then overwriting it with new values; this means that two `beyla.ebpf` components will have a race condition here as they fight over what the "actual" defaults are.

I'd recommend making a copy of beyla.DefaultConfig here.

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (rfratto) [reply docs/sources/flow/reference/components/beyla.ebpf.md:14]: +1 this would be good to mention, as it's not immediately clear to me. 

I think this is best mentioned in the section for the `output` block.

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (rfratto) [inline docs/sources/flow/reference/components/beyla.ebpf.md:40]: nit: your example of a coma-separated list of ports is not a comma-separated list of ports :)

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (rfratto) [inline docs/sources/flow/reference/components/beyla.ebpf.md:87]: nit: for consistency with how we document components, it's a bit weird for all this information to be before the table. Should this be moved down to after the table?

**Review by rfratto — COMMENTED** [review]
Thanks! Mostly looks good, aside from a few concerns in the code.

This primarily will need a review from @clayton-cornell before we can merge, though.

**Comment by rfratto** [comment]
FYI the Mac tests are failing due to doc generation stuff. Ideally that should be fixed before we can merge this

File: internal/component/beyla/beyla_linux.go
Comment (marctc) [reply internal/component/beyla/beyla_linux.go:175]: `/metrics` endpoint will return the RED metrics exposed by Beyla. I followed the prometheus' exporters flow component pattern to implement this, but if there's an easier way to do it, i'm happy to change it.

Comment (rfratto) [reply internal/component/beyla/beyla_linux.go:175]: But are they RED metrics _about_ Beyla? Or about applications Beyla is monitoring?

Comment (marctc) [reply internal/component/beyla/beyla_linux.go]: The RED metrics of the configured applications for Beyla. I'm not exposing internal metrics for now.

Comment (rfratto) [reply internal/component/beyla/beyla_linux.go]: OK 👍 That was the thing I was confused about. If it's metrics for configuration applications, then it makes sense to follow the exporter pattern for those metrics. 

In the future when you get around to exposing internal metrics, those shouldn't be exposed here, and should rather be exposed via the registry passed to the `component.Options` to align with how debug metrics for other components work.

Comment (marctc) [reply internal/component/beyla/beyla_linux.go]: Thanks for the explanation! Should I leave a comment or something?

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (marctc) [reply docs/sources/flow/reference/components/beyla.ebpf.md]: Clarified in the `output` section. @grcevski let me know if it's clear enough

File: internal/component/beyla/beyla_linux.go
Comment (rfratto) [reply internal/component/beyla/beyla_linux.go]: Couldn't hurt!

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:41]:
```suggestion
`open_port` accepts a comma-separated list of ports (for example, `80,443`), and port ranges (for example, `8000-8999`).
If the executable matches only one of the ports in the list, it is considered to match the selection criteria.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:71]:
```suggestion
Allows configuring how some attributes for metrics and traces are decorated.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:110]:
```suggestion
`patterns` and `ignored_patterns` are a list of patterns which a URL path with specific tags which allow for grouping path segments (or ignored them).
The matcher tags can be in the `:name` or `{name}` format.
```

Just a line break

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:111]:
```suggestion
`ignore_mode` properties are:
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:122]:
```suggestion
- `all` discards both metrics and traces that match the `ignored_patterns`.
- `traces` discards only the traces that match the `ignored_patterns`. No metric events are ignored.
- `metrics` discards only the metrics that match the `ignored_patterns`. No trace events are ignored.
`unmatched` properties are:
- `unset` leaves the `http.route` property as unset.
- `path` copies the `http.route` field property to the path value.
  - Caution: this option could lead to a cardinality explosion at the ingester side.
- `wildcard` sets the `http.route` field property to a generic asterisk based `/**` value.
- `heuristic` automatically derives the `http.route` field property from the path value, based on the following rules:
  - Any path components that have numbers or characters outside of the ASCII alphabet (or `-` and _), are replaced by an asterisk `*`.
  - Any alphabetical components that don’t look like words, are replaced by an asterisk `*`.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:134]:
```suggestion
In some scenarios, Beyla will instrument a big variety of services, for example, as a Kubernetes DaemonSet that instruments all the services in a node.
This block allows you to filter the services to instrument based on their metadata.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:143]:
```suggestion
`name` defines a name for the matching instrumented service. It is used to populate the `service.name` OTEL property and/or the `service_name` Prometheus property in the exported metrics/traces.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:144]:
```suggestion
`open_port` accepts a comma-separated list of ports (for example, `80,443`), and port ranges (for example, `8000-8999`). If the executable matches only one of the ports in the list, it is considered to match the selection criteria.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:175]:
```suggestion
[in-memory traffic]: ../../../concepts/component_controller#in-memory-traffic
[run command]: ../../cli/run/
```

We should not be using relref links in any new documentation. The relref short code is deprecated.

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:196]:
```suggestion
    open_port = <OPEN_PORT>
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:210]:
```suggestion
    url = <PROMETHEUS_REMOTE_WRITE_URL>

    basic_auth {
      username = <USERNAME>
      password = <PASSWORD>
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:221]:
```suggestion
- _`<OPEN_PORT>`_: The port of the running service for Beyla automatically instrumented with eBPF.
- _`<PROMETHEUS_REMOTE_WRITE_URL>`_: The URL of the Prometheus remote_write-compatible server to send metrics to.
- _`<USERNAME>`_: The username to use for authentication to the remote_write API.
- _`<PASSWORD>`_: The password to use for authentication to the remote_write API.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:223]:
```suggestion
[scrape]: ../prometheus.scrape/
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:251]:
```suggestion
Replace the following:

- _`<OPEN_PORT>`_: The port of the running service for Beyla automatically instrumented with eBPF.
- _`<OTLP_ENDPOINT>`_: The endpoint of the OpenTelemetry Collector to send traces to.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:231]:
```suggestion
    open_port = <OPEN_PORT>
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:245]:
```suggestion
        endpoint = env("<OTLP_ENDPOINT>")
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:28]:
```suggestion
beyla.ebpf "<LABEL>" {

}
```

**Review by grcevski — APPROVED** [review]
LGTM! On the beyla side of things.

File: internal/component/beyla/ebpf/beyla_linux.go
Comment (rfratto) [inline internal/component/beyla/ebpf/beyla_linux.go:137]: There's nothing setting the health back to healthy, so after this triggers the component will be marked as unhealthy forever

**Review by rfratto — APPROVED** [review]
LGTM after Clayton approves and after the final comment I left is addressed. Thanks Marc!

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:119]:
```suggestion
  - Caution: This option could lead to a cardinality explosion on the ingester side.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:120]:
```suggestion
- `wildcard` sets the `http.route` field property to a generic asterisk-based `/**` value.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:121]:
```suggestion
- `heuristic` automatically derives the `http.route` field property from the path value based on the following rules:
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:123]:
```suggestion
  - Any alphabetical components that don’t look like words are replaced by an asterisk `*`.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:130]:
```suggestion
It contains the following blocks:
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:135]:
```suggestion
In some scenarios, Beyla will instrument a wide variety of services, such as a Kubernetes DaemonSet that instruments all the services in a node.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:152]:
```suggestion
The `output` block configures a set of components to forward the resulting telemetry data to.
```

File: docs/sources/flow/reference/components/beyla.ebpf.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/beyla.ebpf.md:162]:
```suggestion
Configure the `traces` argument to send traces data to other components.
```

**Review by clayton-cornell — COMMENTED** [review]
A few more doc suggestions

**Pull request merged**
